src/tools/perf/cuda/libperf_cuda.h

```diff
@@ -0,0 +1,184 @@
+};
+
+UCS_F_DEVICE ucx_perf_cuda_time_t ucx_perf_cuda_get_time_ns()
+{
+    ucx_perf_cuda_time_t globaltimer;
+    /* 64-bit GPU global nanosecond timer */
+    asm volatile("mov.u64 %0, %globaltimer;" : "=l"(globaltimer));
+    return globaltimer;
+}
+
+UCS_F_DEVICE void
+ucx_perf_cuda_update_report(ucx_perf_cuda_context &ctx,
+                            ucx_perf_counter_t completed,
+                            ucx_perf_counter_t max_iters,
+                            ucx_perf_cuda_time_t &last_report_time)
+{
+    if (threadIdx.x == 0) {
+        ucx_perf_cuda_time_t current_time = ucx_perf_cuda_get_time_ns();
+        if (((current_time - last_report_time) >= ctx.report_interval_ns) ||
+            (completed >= max_iters)) {
+            ctx.completed_iters = completed;
+            last_report_time    = current_time;
+            __threadfence();
+        }
+    }
+}
+
+template <typename T>
+class ucx_perf_cuda_device_mem {
+public:
+    ucx_perf_cuda_device_mem(size_t size = sizeof(T))
+    {
+        ucs_status_t status = ucx_perf_cuda_device_mem_create(&m_mem, size);
+        if (status != UCS_OK) {
+            ucs_fatal("ucx_perf_cuda_device_mem_create() failed: %s",
+                      ucs_status_string(status));
+        }
+    }
+
+    ~ucx_perf_cuda_device_mem()
+    {
+        ucx_perf_cuda_device_mem_destroy(&m_mem);
+    }
+
+    T *cpu_ptr() const { return static_cast<T*>(m_mem.cpu_ptr); }
+    T *gpu_ptr() const { return static_cast<T*>(m_mem.gpu_ptr); }
+
+private:
+    ucx_perf_cuda_device_mem_t m_mem;
+};
+
+template <typename T>
+class ucx_perf_cuda_mem {
+public:
+    ucx_perf_cuda_mem(size_t size = sizeof(T))
+    {
+        CUDA_CALL_HANDLER(ucs_fatal, , cudaMalloc, &m_mem, size);
+    }
+
+    ~ucx_perf_cuda_mem()
+    {
+        CUDA_CALL_HANDLER(ucs_warn, , cudaFree, m_mem);
+    }
+
+    T *ptr() const { return static_cast<T*>(m_mem); }
+
+private:
+    void *m_mem;
+};
+
+template <typename Base>
+class ucx_perf_cuda_test_runner: public Base {
+public:
+    using psn_t = uint64_t;
+    using Base::m_perf;
+
+    ucx_perf_cuda_test_runner(ucx_perf_context_t &perf) : Base(perf)
+    {
+        ucx_perf_cuda_context *cpu_ctx = m_device_mem.cpu_ptr();
+
+        cpu_ctx->max_outstanding       = perf.params.max_outstanding;
+        cpu_ctx->max_iters             = perf.params.max_iter;
+        cpu_ctx->report_interval_ns    = perf.params.report_interval *
+                                         UCS_NSEC_PER_SEC;
+        cpu_ctx->completed_iters       = 0;
+
+        m_poll_interval                = perf.params.report_interval / 10000;
+        m_index                        = rte_call(&perf, group_index);
+        m_mem_type                     = m_index ? perf.params.send_mem_type :
+                                                   perf.params.recv_mem_type;
+        m_allocator                    = m_index ? perf.send_allocator :
+                                                   perf.recv_allocator;
```

Comment:
nextpr: try to remove this vars and if critical use helper inline function, since it's all slow path and not part of the kernel